File: skills/code-architect/references/code-examples/php/Strategy.php
<?php
// Strategy — a family of interchangeable algorithms behind one interface.
declare(strict_types=1);

interface SortStrategy
{
    /**
     * @param int[] $items
     * @return int[]
     */
    public function sort(array $items): array;
}

final class AscendingStrategy implements SortStrategy
{
    public function sort(array $items): array
    {
        sort($items);
        return $items;
    }
}

final class DescendingStrategy implements SortStrategy
{
    public function sort(array $items): array
    {
        rsort($items);
        return $items;
    }
}

final class ByAbsoluteStrategy implements SortStrategy
{
    public function sort(array $items): array
    {
        usort($items, static fn (int $a, int $b) => abs($a) <=> abs($b));
        return $items;
    }
}

// Context delegates the algorithm to the injected strategy.
final class Sorter
{
    public function __construct(private SortStrategy $strategy) {}

    public function setStrategy(SortStrategy $strategy): void
    {
        $this->strategy = $strategy;
    }

    /**
     * @param int[] $items
     * @return int[]
     */
    public function run(array $items): array
    {
        return $this->strategy->sort($items);
    }
}

// Demo
$data = [3, -1, 2, -5];
$sorter = new Sorter(new AscendingStrategy());
echo implode(',', $sorter->run($data)) . PHP_EOL;
$sorter->setStrategy(new ByAbsoluteStrategy());
echo implode(',', $sorter->run($data)) . PHP_EOL;

File: skills/code-architect/references/code-examples/rust/strategy.rs
// Strategy — define a family of interchangeable algorithms behind one interface
// so the algorithm can be selected and swapped at runtime.

trait CompressionStrategy {
    fn compress(&self, data: &str) -> String;
}

struct RunLength;
struct Doubling;

impl CompressionStrategy for RunLength {
    fn compress(&self, data: &str) -> String {
        let mut out = String::new();
        let mut chars = data.chars().peekable();
        while let Some(c) = chars.next() {
            let mut count = 1;
            while chars.peek() == Some(&c) {
                chars.next();
                count += 1;
            }
            out.push_str(&format!("{}{}", count, c));
        }
        out
    }
}

impl CompressionStrategy for Doubling {
    fn compress(&self, data: &str) -> String {
        data.chars().flat_map(|c| [c, c]).collect()
    }
}

// Context delegates the work to whichever strategy it currently holds.
struct Archiver {
    strategy: Box<dyn CompressionStrategy>,
}

impl Archiver {
    fn run(&self, data: &str) -> String {
        self.strategy.compress(data)
    }
}

fn main() {
    let mut archiver = Archiver { strategy: Box::new(RunLength) };
    println!("rle: {}", archiver.run("aaabbc"));

    archiver.strategy = Box::new(Doubling);
    println!("doubling: {}", archiver.run("abc"));
}
